### PR TITLE
Improved copy/pasting in ExtendedTableWidget

### PR DESCRIPTION
**Pull request by innermous**
* NULLs are now copied and pasted correctly
* Ctrl+C allows only rectangular selections (multi-selections should be provided later with Ctrl+Shift+C and not use clipboard)

Testing is much appreciated.

NB. return in 199L is workaround, because  of mess with keyPressEvent in 226L.

### CODE REVIEW

**Comment by revolter** [comment]
Should it work for BLOBs too? Also, I found a problem, but can't easily reproduce. I had a database which got corrupted because of #674, and had the content

```
"�PNG

"
```

upon copying it and trying to paste in a cell, it pastes `"�PNG` __and__ `"` in the cell beneath it, even though I copied only one cell.

Running `sqlite3 db.sqlite "SELECT column2 FROM table WHERE column1 = '1';" | hexdump -c` gives me:

```
0000000   �   �   �   P   N   G  \r  \n 032  \n  \n
000000b
```

and I believe the `\r` is the problem, but I can't manage to type that character manually on OS X.

**Comment by revolter** [comment]
Finally reproduced it using (while learning) the hex editor.

- select a cell
- switch to Binary mode in the edit dock
- type `31 0d 0a 1a 32`
- copy & paste the cell

and it will paste `"1` in the target cell and `2"` (actually `1a 32 22`, where `1a` is `\r`) in the cell beneath it.

**Comment by justinclift** [comment]
Excellent @innermous, this looks like a good start to improving the Copy/Paste code. 👍 

* The new warning message about "This function cannot be used with multiple selections." is a decent idea.  Maybe add the word "independent" to it though? eg "... with multiple independent selections."

  Without that, it could be kind of confusing.  Some people might misunderstand it to mean "multiple cells selected".

* The code in this PR is making progress towards better BLOB handing too.  Prior to this PR, binary data ("BLOBS") would just copy the text string "BLOB" to a target cell.  With this PR, at least it's attempting to copy the actual data. 😁 

  The copy itself doesn't seem binary safe though, as @revolter mentioned above.  In testing here, it's converting image blobs to corrupted text (eg "PNG") and also corrupting the cell below it.

  The screenshots below should illustrate.

  screen shot 2016-08-06 at 14 18 34
  screen shot 2016-08-06 at 14 18 11

**Comment by innermous** [comment]
Wait, copying blobs works without my commit?

**Comment by revolter** [comment]
@innermous, That's what @justinclift said:

> Prior to this PR, binary data ("BLOBS") would just copy the text string "BLOB" to a target cell.

**Comment by justinclift** [comment]
Yeah, without this PR it just copies the text "BLOB" without the real data. :wink:

File: src/ExtendedTableWidget.cpp [hunk 1]
```diff
@@ -28,60 +28,131 @@ ExtendedTableWidget::ExtendedTableWidget(QWidget* parent) :
...
+            containsBinary = true;
+            break;
+        }
+
+    if (containsBinary) {
+        qApp->clipboard()->clear();
+        // Copy selected data into inner buffer
+        int columns = indices.last().column() - indices.first().column() + 1;
+        while (!indices.isEmpty()) {
+            QByteArrayList lst;
```
Comment (justinclift) [inline src/ExtendedTableWidget.cpp]: Seems to need an `#include <QByteArrayList>` up top, as Travis is complaining.

Comment (innermous) [inline src/ExtendedTableWidget.cpp]: Never mind, it'll all be fixed.
It's not even ready.

I am astonished though, why my build didn't fail.

**Comment by innermous** [comment]
I don't get it.
I included <QByteArrayList>, but it's still mad.

File: src/ExtendedTableWidget.cpp (hunk 1 above)
Comment (innermous) [inline src/ExtendedTableWidget.cpp]: Well, I got it.
They don't have QByteArrayList in Qt4. Another QByteArrayList in SQLiteBrowser code is just typedef.
You still think we shouldn't move to Qt5?

**Comment by justinclift** [comment]
Yeah, we definitely can't move to Qt5 yet. 😦 

I'll build our OSX packages for 3.9.0 with Qt4 again for this release.  If they work ok for people, and we don't get crash reports (like last time, with 3.8.0), we can probably move to Qt5 only. :smile:

**Comment by revolter** [comment]
Looking good. BLOBs are working now, but [this][1] still corrupts the cell beneath the target cell.

  [1]: https://github.com/sqlitebrowser/sqlitebrowser/pull/709#issuecomment-238015251

**Comment by innermous** [comment]
@revolter I don't know how come it got corrupted for you, since multi-blob data is managed via inner buffer.

**Comment by justinclift** [comment]
Hmmm, I broke the "Apply" button in the Edit Cell with yesterday's change.  I need to fix that first, in order to test some stuff properly.

Looking at it now. :smile:

**Comment by revolter** [comment]
@innermous, I already explained it, please read again. It's not about copying multiple cells.

**Comment by innermous** [comment]
@revolter can you try with latest commit and no corrupted data in database? I can't really reproduce it, I tried countless images, they all work correctly.

**Comment by justinclift** [comment]
Closed?

**Comment by innermous** [comment]
@justinclift I don't know how to implement multilines correctly.
Maybe in future I'll figure out but for now I'll freeze the branch.

**Comment by revolter** [comment]
@innermous, Didn't you say that the multiline problem existed prior to this PR? I think that this adds useful fixed like copying BLOBs correctly.

**Comment by justinclift** [comment]
Thinking about it, @revolter is correct.  This PR fixes some of the existing broken behaviour (binary/image copying internally), without introducing new bugs.

It's not a "perfect" solution, but it's better than the code we have.  Are you ok with it being merged anyway? :smile:

**Comment by innermous** [comment]
Resuming work.

**Comment by justinclift** [comment]
@innermous Ok. :smile:

An idea for this popped into my head last night.

For the internal copy-paste operation, would it be possible to treat text type cells as binary?

eg:

* isBinary() continues to put text data on the clipboard for sharing-between-applications
* text data is also added to m_buffer (maybe as QByteArray too?)
* for paste operation, m_buffer is used unless external data has been added to clipboard. (for example, someone does Copy in text editor then Paste in DB4S)

Detecting external copy to clipboard can probably be done using checksum created for Copy operation, and maybe kept in a class variable

File: src/ExtendedTableWidget.cpp [hunk 2]
```diff
@@ -28,59 +29,153 @@ ExtendedTableWidget::ExtendedTableWidget(QWidget* parent) :
...
+    if (indices.size() == 1) {
+        QImage img;
+        if (img.loadFromData(m->data(indices.first(), Qt::EditRole).toByteArray())) {
+            qApp->clipboard()->setImage(img);
+            return;
+        }
+    }
+
+    // Check whether selection is rectangular
+    QItemSelection rect(indices.first(), indices.last());
```
Comment (justinclift) [inline src/ExtendedTableWidget.cpp]: Would setting `QAbstractItemView::ContiguousSelection` instead, to force a single continuous selection, be better?

**Comment by innermous** [comment]
Moved to google docs format. 
I need to test it on windows first, it has some dibs with carriage return.

File: src/ExtendedTableWidget.cpp (hunk 2 above)
Comment (innermous) [inline src/ExtendedTableWidget.cpp]: If we are not implementing multi-selections copy then you can force rectangular selection.

**Comment by innermous** [comment]
Feeling sorry for your travis.
Adding some windows-specific changes required to go deep in history :)

**Comment by justinclift** [comment]
Awesome. :smile:

**Comment by justinclift** [comment]
Testing this out now on OSX, it seems pretty good.

Copying & pasting BLOB data works properly, as does multi-line text, as does cut-n-paste to external applications.

I'm going to cherry-pick this into the 3.9.x branch too. :wink:

**Pull request merged**